lib/grandpa/finalisation.go

```diff
@@ -0,0 +1,467 @@
+	servicesLock       sync.Mutex
+	finalisationEngine ephemeralService
+	votingRound        ephemeralService
+
+	// newServices is a constructor function which takes care to instantiate
+	// and return the services needed to finalize a round, those services
+	// are ephemeral services with a lifetime of a round
+	newServices   func() (engine, voting ephemeralService)
+	initiateRound func() error
+
+	stopCh      chan struct{}
+	handlerDone chan struct{}
+}
+
+func newFinalisationHandler(service *Service) *finalisationHandler {
+	return &finalisationHandler{
+		newServices: func() (engine, voting ephemeralService) {
+			finalisationEngine := newfinalisationEngine(service)
+			votingRound := newvotingRoundHandler(service, finalisationEngine.actionCh)
+			return finalisationEngine, votingRound
+		},
+		initiateRound: service.initiateRound,
+		stopCh:        make(chan struct{}),
+		handlerDone:   make(chan struct{}),
+	}
+}
+
+func (fh *finalisationHandler) Start() (<-chan error, error) {
+	errorCh := make(chan error)
+	go fh.run(errorCh)
+
+	return errorCh, nil
+}
+
+func (fh *finalisationHandler) run(errorCh chan<- error) {
+	defer func() {
+		close(errorCh)
+		close(fh.handlerDone)
+	}()
+
+	for {
+		select {
+		case <-fh.stopCh:
+			return
+		default:
+		}
+
+		err := fh.initiateRound()
+		if err != nil {
+			errorCh <- fmt.Errorf("initiating round: %w", err)
+			return
+		}
+
+		err = fh.runEphemeralServices()
+		if err != nil {
+			errorCh <- fmt.Errorf("running ephemeral services: %w", err)
+			return
+		}
+	}
+}
+
+func (fh *finalisationHandler) stop() (err error) {
+	fh.servicesLock.Lock()
+	defer fh.servicesLock.Unlock()
+
+	finalisationEngineErrCh := make(chan error)
+	go func() {
+		finalisationEngineErrCh <- fh.finalisationEngine.Stop()
+	}()
+
+	votingRoundErrCh := make(chan error)
+	go func() {
+		votingRoundErrCh <- fh.votingRound.Stop()
+	}()
+
+	finalisationEngErr := <-finalisationEngineErrCh
+	votingRoundErr := <-votingRoundErrCh
+
+	switch {
+	case finalisationEngErr != nil && votingRoundErr != nil:
+		return fmt.Errorf("%w: %s; %s", errServicesStopFailed, finalisationEngErr, votingRoundErr)
+	case finalisationEngErr != nil:
+		return fmt.Errorf("%w: %s", errServicesStopFailed, finalisationEngErr)
+	case votingRoundErr != nil:
+		return fmt.Errorf("%w: %s", errServicesStopFailed, votingRoundErr)
+	}
+
+	return nil
+}
+
+func (fh *finalisationHandler) Stop() (err error) {
+	close(fh.stopCh)
+	<-fh.handlerDone
+
+	return fh.stop()
+}
+
+// runEphemeralServices starts the two ephemeral services that handle the
+// votes for the current round, and returns with nil when the two
+// service runs succeed.
+// If any service run fails, the other service run is stopped and
+// an error is returned. The function returns nil is the finalisation
+// handler is stopped.
+func (fh *finalisationHandler) runEphemeralServices() error {
+	fh.servicesLock.Lock()
+	fh.finalisationEngine, fh.votingRound = fh.newServices()
+	fh.servicesLock.Unlock()
```

Comment:
Start() should not return until these are set to something non-nil, otherwise further calls like `.Stop()` would panic on nil pointers. See the related comment on finalisation_integration_test.go